File: ui/src/command_palette_match.rs
//! Pure matching/ranking logic behind the command palette: the searchable
//! `Command` model, fuzzy scoring, ranking, the records→`Command` builder, and
//! the keyboard selection-index helpers. Split out of `command_palette.rs` to
//! keep that file under the line-count gate; the component in that file is a
//! thin shell over these host-testable functions (see
//! `command_palette_tests.rs`).

use crate::routines::Routine;

/// What a palette entry points at — a fixed page, or an entity that lives on a
/// page. Kept free of `yew_router::Route` so the ranking/build logic is fully
/// host-testable; the component maps each variant to a concrete route.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub(crate) enum CmdKind {
    /// Jump to the OVERVIEW page.
    NavOverview,
    /// Jump to the ROUTINES page.
    NavRoutines,
    /// Jump to the HEATMAP page.
    NavHeatmap,
    /// Jump to the SETTINGS page.
    NavSettings,
    /// A specific routine (lands on the ROUTINES page).
    Routine,
    /// Re-poll server health (the header's ↻ action).
    ActionRefresh,
    /// Open the stop-server confirmation (the header's ⏻ action).
    ActionStop,
    /// Toggle the light/dark theme (persisted to localStorage).
    ActionToggleTheme,
}

/// The destination page a [`CmdKind`] resolves to, independent of the wasm
/// router so it can be asserted on the host.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub(crate) enum RouteKind {
    /// The OVERVIEW page (`/`).
    Home,
    /// The ROUTINES page (`/routines`).
    Routines,
    /// The HEATMAP page (`/heatmap`).
    Heatmap,
    /// The SETTINGS page (`/settings`).
    Settings,
}

/// One searchable destination in the palette.
#[derive(Clone, PartialEq, Eq, Debug)]
pub(crate) struct Command {
    /// What this entry points at.
    pub kind: CmdKind,
    /// Primary label shown and matched first.
    pub title: String,
    /// Secondary line (schedule description / route hint).
    pub subtitle: String,
    /// Extra terms folded into the fuzzy haystack (agent, handler, id, …) so a
    /// match on an alias still surfaces the entry.
    pub keywords: String,
}

/// The page a command navigates to, or `None` for an action command (which
/// runs a callback instead of navigating).
pub(crate) fn route_for(kind: CmdKind) -> Option<RouteKind> {
    match kind {
        CmdKind::NavOverview => Some(RouteKind::Home),
        CmdKind::NavRoutines | CmdKind::Routine => Some(RouteKind::Routines),
        CmdKind::NavHeatmap => Some(RouteKind::Heatmap),
        CmdKind::NavSettings => Some(RouteKind::Settings),
        CmdKind::ActionRefresh | CmdKind::ActionStop | CmdKind::ActionToggleTheme => None,
    }
}

/// The short category badge text for a command (used in the row and as the
/// group separator label).
pub(crate) fn badge_for(kind: CmdKind) -> &'static str {
    match kind {
        CmdKind::NavOverview
        | CmdKind::NavRoutines
        | CmdKind::NavHeatmap
        | CmdKind::NavSettings => "GO",
        CmdKind::Routine => "ROUTINE",
        CmdKind::ActionRefresh | CmdKind::ActionStop | CmdKind::ActionToggleTheme => "ACTION",
    }
}

/// Score how well `query` fuzzy-matches `text`: both are compared
/// case-insensitively, and `query` must appear as an ordered subsequence of
/// `text`. A higher score is a better match. Returns `None` when `query` is not
/// a subsequence; an empty (or whitespace-only) query matches everything with a
/// neutral score of `0`, so the unfiltered list keeps its natural order.
///
/// Bonuses reward the matches users perceive as "good": a hit at the very start
/// of the text, a hit right after a word boundary, and runs of consecutive
/// characters. Longer texts are mildly penalized so a tight match on a short
/// label outranks a scattered one on a long string.
pub(crate) fn fuzzy_score(text: &str, query: &str) -> Option<i32> {
    let needle = query.trim();
    if needle.is_empty() {
        return Some(0);
    }
    let hay: Vec<char> = text.to_lowercase().chars().collect();
    let pins: Vec<char> = needle.to_lowercase().chars().collect();
    let mut score = 0i32;
    let mut cursor = 0usize;
    let mut prev: Option<usize> = None;
    for &needle_ch in &pins {
        let mut hit = None;
        while cursor < hay.len() {
            let hay_ch = hay[cursor];
            cursor += 1;
            if hay_ch == needle_ch {
                hit = Some(cursor - 1);
                break;
            }
        }
        let pos = hit?;
        score += 1;
        if pos == 0 {
            score += 10; // start of string
        } else if !hay[pos - 1].is_alphanumeric() {
            score += 6; // start of a word
        }
        if let Some(prev_pos) = prev {
            if pos == prev_pos + 1 {
                score += 8; // consecutive run
            }
        }
        prev = Some(pos);
    }
    score -= hay.len() as i32 / 16;
    Some(score)
}

/// Best fuzzy score for a command against `query`, matching the title at full
/// weight and the keyword aliases at a slight discount so a title hit wins a
/// tie. `None` when neither field matches.
fn command_score(command: &Command, query: &str) -> Option<i32> {
    let title = fuzzy_score(&command.title, query);
    let alias = fuzzy_score(&command.keywords, query).map(|raw| raw - 4);
    match (title, alias) {
        (Some(left), Some(right)) => Some(left.max(right)),
        (left, right) => left.or(right),
    }
}

/// Indices of `commands` that match `query`, best-first. Ties keep the input
/// order (which is already grouped: pages, then routines), so an empty query
/// returns every command in its natural grouping.
pub(crate) fn rank(commands: &[Command], query: &str) -> Vec<usize> {
    let mut scored: Vec<(usize, i32)> = commands
        .iter()
        .enumerate()
        .filter_map(|(idx, command)| command_score(command, query).map(|score| (idx, score)))
        .collect();
    scored.sort_by(|left, right| right.1.cmp(&left.1).then_with(|| left.0.cmp(&right.0)));
    scored.into_iter().map(|(idx, _)| idx).collect()
}

/// Build the full command list: the pages first, then one entry per routine.
/// Subtitles prefer the server's human schedule description and fall back to
/// the raw expression.
pub(crate) fn build_commands(routines: &[Routine]) -> Vec<Command> {
    let mut commands = vec![
        Command {
            kind: CmdKind::NavOverview,
            title: "Overview".into(),
            subtitle: "Fleet summary & upcoming runs".into(),
            keywords: "home dashboard kpi summary landing".into(),
        },
        Command {
            kind: CmdKind::NavRoutines,
            title: "Routines".into(),
            subtitle: "Manage agent-driven routines".into(),
            keywords: "agents automation".into(),
        },
        Command {
            kind: CmdKind::NavHeatmap,
            title: "Heatmap".into(),
            subtitle: "7-day × 24-hour fire-density grid".into(),
            keywords: "schedule density grid busy collisions calendar".into(),
        },
        Command {
            kind: CmdKind::NavSettings,
            title: "Settings".into(),
            subtitle: "Persistent agent prompt".into(),
            keywords: "config preferences user prompt".into(),
        },
        Command {
            kind: CmdKind::ActionRefresh,
            title: "Refresh".into(),
            subtitle: "Re-poll server health".into(),
            keywords: "reload health status action".into(),
        },
        Command {
            kind: CmdKind::ActionStop,
            title: "Stop Server".into(),
            subtitle: "Shut the moadim server down".into(),
            keywords: "shutdown halt kill quit action".into(),
        },
        Command {
            kind: CmdKind::ActionToggleTheme,
            title: "Toggle Theme".into(),
            subtitle: "Switch between dark and light mode".into(),
            keywords: "theme light dark mode toggle appearance action".into(),
        },
    ];
    for routine in routines {
        commands.push(Command {
            kind: CmdKind::Routine,
            title: routine.title.clone(),
            subtitle: routine_subtitle(routine),
            keywords: format!(
                "{} {} {} {} routine",
                routine.id,
                routine.agent,
                routine.schedule,
                routine.tags.join(" ")
            ),
        });
    }
    commands
}

/// Subtitle for a routine command: schedule label optionally suffixed with
/// comma-separated status tags so health issues are visible in search results.
pub(crate) fn routine_subtitle(routine: &Routine) -> String {
    let sched = schedule_label(routine.schedule_description.as_ref(), &routine.schedule);
    let mut tags: Vec<&str> = Vec::new();
    if !routine.enabled {
        tags.push("DISABLED");
    } else if routine.skip_runs.is_some_and(|n| n > 0) {
        tags.push("SNOOZED");
    } else if !routine.agent_registered {
        tags.push("AGENT MISSING");
    }
    if routine.flag_count > 0 {
        tags.push("FLAGS");
    }
    if tags.is_empty() {
        sched
    } else {
        format!("{sched} — {}", tags.join(", "))
    }
}

/// The human schedule description when present, else the raw expression, else a
/// dash so a row never renders an empty subtitle.
pub(crate) fn schedule_label(human: Option<&String>, raw: &str) -> String {
    match human {
        Some(text) if !text.is_empty() => text.clone(),
        _ if !raw.trim().is_empty() => raw.to_string(),
        _ => "—".into(),
    }
}

/// Clamp `selected` to a valid row index for a result list of `len` rows: stays
/// at `0` when empty, otherwise never points past the last row.
pub(crate) fn clamp_selection(selected: usize, len: usize) -> usize {
    if len == 0 {
        0
    } else {
        selected.min(len - 1)
    }
}

/// Next selection index when pressing ↓: advances by one but never past the
/// last row (no wrap), and stays at `0` for an empty list.
pub(crate) fn next_index(selected: usize, len: usize) -> usize {
    if len == 0 {
        0
    } else {
        (selected + 1).min(len - 1)
    }
}

/// Previous selection index when pressing ↑: retreats by one, saturating at the
/// first row (no wrap).
pub(crate) fn prev_index(selected: usize) -> usize {
    selected.saturating_sub(1)
}

/// Last selection index when pressing End: the final row, or `0` when empty.
pub(crate) fn last_index(len: usize) -> usize {
    len.saturating_sub(1)
}

File: client/src/shell/commandPaletteMatch.ts
import type { RoutineResponse } from "../api/hooks";

/** What a palette entry points at. */
export type CmdKind =
  | "nav-overview"
  | "nav-routines"
  | "nav-heatmap"
  | "nav-reliability"
  | "nav-machines"
  | "nav-settings"
  | "routine"
  | "action-refresh"
  | "action-stop"
  | "action-toggle-theme";

export type RouteKind = "home" | "routines" | "heatmap" | "reliability" | "machines" | "settings";

export interface Command {
  kind: CmdKind;
  title: string;
  subtitle: string;
  /** Extra terms folded into the fuzzy haystack (agent, schedule, id, tags…). */
  keywords: string;
  /** Present only for `kind: "routine"` commands. */
  routineId?: string;
}

/** The page a command navigates to, or `undefined` for an action command. */
export function routeFor(kind: CmdKind): RouteKind | undefined {
  switch (kind) {
    case "nav-overview":
      return "home";
    case "nav-routines":
    case "routine":
      return "routines";
    case "nav-heatmap":
      return "heatmap";
    case "nav-reliability":
      return "reliability";
    case "nav-machines":
      return "machines";
    case "nav-settings":
      return "settings";
    default:
      return undefined;
  }
}

/** Short category badge text for a command. */
export function badgeFor(kind: CmdKind): string {
  switch (kind) {
    case "nav-overview":
    case "nav-routines":
    case "nav-heatmap":
    case "nav-reliability":
    case "nav-machines":
    case "nav-settings":
      return "GO";
    case "routine":
      return "ROUTINE";
    default:
      return "ACTION";
  }
}

/**
 * Fuzzy-scores `query` as an ordered subsequence of `text` (case-insensitive).
 * Returns `undefined` when `query` isn't a subsequence of `text`; a blank
 * query matches everything with a neutral score of `0`. Bonuses reward a hit
 * at the start of the text, a hit after a word boundary, and consecutive
 * runs; longer texts are mildly penalized.
 */
export function fuzzyScore(text: string, query: string): number | undefined {
  const needle = query.trim();
  if (needle === "") return 0;
  const hay = [...text.toLowerCase()];
  const pins = [...needle.toLowerCase()];
  let score = 0;
  let cursor = 0;
  let prev: number | undefined;
  for (const needleCh of pins) {
    let hit: number | undefined;
    while (cursor < hay.length) {
      const hayCh = hay[cursor];
      cursor += 1;
      if (hayCh === needleCh) {
        hit = cursor - 1;
        break;
      }
    }
    if (hit === undefined) return undefined;
    score += 1;
    if (hit === 0) {
      score += 10;
    } else if (!/[a-z0-9]/i.test(hay[hit - 1] ?? "")) {
      score += 6;
    }
    if (prev !== undefined && hit === prev + 1) {
      score += 8;
    }
    prev = hit;
  }
  score -= Math.floor(hay.length / 16);
  return score;
}

function commandScore(command: Command, query: string): number | undefined {
  const title = fuzzyScore(command.title, query);
  const aliasRaw = fuzzyScore(command.keywords, query);
  const alias = aliasRaw === undefined ? undefined : aliasRaw - 4;
  if (title !== undefined && alias !== undefined) return Math.max(title, alias);
  return title ?? alias;
}

/** Indices of `commands` matching `query`, best-first; ties keep input order. */
export function rank(commands: Command[], query: string): number[] {
  const scored: Array<[number, number]> = [];
  commands.forEach((command, idx) => {
    const score = commandScore(command, query);
    if (score !== undefined) scored.push([idx, score]);
  });
  scored.sort(([aIdx, aScore], [bIdx, bScore]) => bScore - aScore || aIdx - bIdx);
  return scored.map(([idx]) => idx);
}

/** The human schedule description when present, else the raw expression, else a dash. */
export function scheduleLabel(human: string | null | undefined, raw: string): string {
  if (human) return human;
  return raw.trim() !== "" ? raw : "—";
}

/** Subtitle for a routine command: schedule label plus status tags, if any. */
export function routineSubtitle(routine: RoutineResponse): string {
  const sched = scheduleLabel(routine.schedule_description, routine.schedule);
  const tags: string[] = [];
  if (!routine.enabled) {
    tags.push("DISABLED");
  } else if ((routine.skip_runs ?? 0) > 0) {
    tags.push("SNOOZED");
  } else if (!routine.agent_registered) {
    tags.push("AGENT MISSING");
  }
  if (routine.flag_count > 0) tags.push("FLAGS");
  return tags.length === 0 ? sched : `${sched} — ${tags.join(", ")}`;
}

/** Builds the full command list: pages first, then one entry per routine. */
export function buildCommands(routines: RoutineResponse[]): Command[] {
  const commands: Command[] = [
    {
      kind: "nav-overview",
      title: "Overview",
      subtitle: "Fleet summary & upcoming runs",
      keywords: "home dashboard kpi summary landing",
    },
    {
      kind: "nav-routines",
      title: "Routines",
      subtitle: "Manage agent-driven routines",
      keywords: "agents automation",
    },
    {
      kind: "nav-heatmap",
      title: "Heatmap",
      subtitle: "7-day x 24-hour fire-density grid",
      keywords: "schedule density grid busy collisions calendar",
    },
    {
      kind: "nav-reliability",
      title: "Reliability",
      subtitle: "Success rate, streaks & duration regressions",
      keywords: "flaky failing streak p50 p95 duration regression health",
    },
    {
      kind: "nav-machines",
      title: "Machines",
      subtitle: "Per-host fleet inventory",
      keywords: "fleet hosts inventory nodes servers",
    },
    {
      kind: "nav-settings",
      title: "Settings",
      subtitle: "Persistent agent prompt",
      keywords: "config preferences user prompt",
    },
    {
      kind: "action-refresh",
      title: "Refresh",
      subtitle: "Re-poll server health",
      keywords: "reload health status action",
    },
    {
      kind: "action-stop",
      title: "Stop Server",
      subtitle: "Shut the moadim server down",
      keywords: "shutdown halt kill quit action",
    },
    {
      kind: "action-toggle-theme",
      title: "Toggle Theme",
      subtitle: "Switch between dark and light mode",
      keywords: "theme light dark mode toggle appearance action",
    },
  ];
  for (const routine of routines) {
    commands.push({
      kind: "routine",
      title: routine.title,
      subtitle: routineSubtitle(routine),
      keywords: `${routine.id} ${routine.agent} ${routine.schedule} ${(routine.tags ?? []).join(" ")} routine`,
      routineId: routine.id,
    });
  }
  return commands;
}

/** Clamps `selected` to a valid row index for a result list of `len` rows. */
export function clampSelection(selected: number, len: number): number {
  return len === 0 ? 0 : Math.min(selected, len - 1);
}

/** Next selection index (↓): advances by one, never past the last row (no wrap). */
export function nextIndex(selected: number, len: number): number {
  return len === 0 ? 0 : Math.min(selected + 1, len - 1);
}

/** Previous selection index (↑): retreats by one, saturating at the first row (no wrap). */
export function prevIndex(selected: number): number {
  return Math.max(0, selected - 1);
}

/** Last selection index (End): the final row, or `0` when empty. */
export function lastIndex(len: number): number {
  return Math.max(0, len - 1);
}
